---
title: "560 HW #1"
author: "David Bacsik"
date: "April 10, 2018"

output: html_document
---
### *due date: Thursday, April 12, 2018*

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Part 1
#### 1.1a
Load the master data file and select the data I want from it. I am going to use the ww_data we used in the lab.

```{r}
master_data = read.table(file="560_dataset_reduced.txt", header = T, sep = '\t')

head(master_data)

ww_data = master_data[which(master_data$dms_id == 'WW_domain'),]

head(ww_data)
```

#### 1.1b
Now, I want to calculate some summary statistics about the functional scores of each variant.

```{r}
func_scores = ww_data$reported_effect


fs.range = range(func_scores)

fs.mean = mean(func_scores)

fs.var = var(func_scores)

fs.sd = sd(func_scores)

quant_list = c(0.025, 0.05, 0.25, 0.5, 0.75, 0.95, 0.975)
fs.quants = quantile(func_scores, probs=quant_list)
```

The range of functional scores is: `r fs.range`

The mean of functional scores is: `r fs.mean`

The variance of functional scores is: `r fs.var`

The standard deviation of functional scores is: `r fs.sd`

Quantiles:

```{r} 
print(fs.quants)
```

Are the 25% and 75% quantiles approximately .67 sd from the mean?'

```{r}
approx_25 = mean(func_scores)-0.67448*sd(func_scores)
approx_75 = mean(func_scores)+0.67488*sd(func_scores)
```

In a normal distribution, the 25th percentile would be approximately `r approx_25`, and in our data it is approximately `r quantile(func_scores, probs=c(0.25))`. The 75th percentile woudl be approximately `r approx_75`, and in our data it is approximately `r quantile(func_scores, probs=c(0.75))`.

These numbers are pretty close, so based on this, I conclude that our data are approximately normally distributed.

#### 1.1c
I want to visualize the distribution of our data. To do this, I will make a histogram of the functional scores with 50 bins.

```{r}

hist(func_scores, 50, freq=TRUE, main='WW DMS Functional Scores', xlab='Score', ylab='Frequency')
```

#### 1.1d
I'd like to see if there are broad trends to mutation to a specific amino acid. For each of the 20 amino acids, I will calculate the mean effect of mutation to that amino acid.

```{r}

effect_by_aa.mean = tapply(ww_data$reported_effect, ww_data$aa2, mean)

effect_by_aa.mean

```

#### 1.1e
Now I will plot these data on a barplot.

```{r}

barplot(effect_by_aa.mean, main = 'WW DMS Substituted Amino Acids', xlab = 'Amino Acid', ylab = 'Mean Functional Score', cex.names=0.8)

```

#### 1.1f
Focusing just on stop codons, alaline, and tryptophan, I will plot the distribution of functional scores for each of these amino acid substitutions.

```{r}

effect_by_aa = split(ww_data$reported_effect, ww_data$aa2)

sub = c(effect_by_aa['X'], effect_by_aa['A'], effect_by_aa['W'])

boxplot(sub, main = 'WW DMS Substituted Amino Acids', names = c('Stop','Alanine','Tryptophan'), xlab= 'Amino Acid', ylab='Functional Scores')
```

#### 1.1g
Now, I would like to see how the above distributions of scores compares to a random sampling of scores.

First, I will pool all the scores for these three codons. Then, I will generate a random sample of the same size. Finally, I will see if the real scores trend above, below, or similarly to the randmon sampling to determine whether they are random or not.

Because I am randomly sampling, and I want this to be repeatable, I will set the seed for this chunk of code at 1.

```{r}
set.seed(1)

combined_scores = unlist(c(effect_by_aa['X'], effect_by_aa['A'], effect_by_aa['W']), use.names=FALSE)

random_sample.x = sample(combined_scores, length(unlist(effect_by_aa['X'])))
random_sample.a = sample(combined_scores, length(unlist(effect_by_aa['A'])))
random_sample.w = sample(combined_scores, length(unlist(effect_by_aa['W'])))
random_sample = list(X = random_sample.x, A = random_sample.a, W = random_sample.w)

boxplot(random_sample, main = 'WW DMS Randomized Amino Acid Substitution', names = c('Randomized Stop','Random Alanine','Randomized Tryptophan'), xlab= 'Amino Acid (Randomized)', ylab='Randomized Functional Scores')
```

The random distribution is pretty uniform, while the real data lie above (maybe) and below (certainly) the uniform distribtion. I am conlcuding that stop codon substituions have a real, negative effect on function; alanine and tryptophan substitutions may be slightly less deleterious than a random substitution.

### Part 2
